typings/index.d.ts

```diff
@@ -0,0 +1,229 @@
+  GuildMember,
+  Webhook,
+  StringResolvable,
+  APIMessage,
+  Message,
+  WebhookClient
+} from "discord.js";
+
+export function main(client:Client): void;
+
+export type ApplicationCommandTypes = "CHAT_INPUT" | "USER" | "MESSAGE";
+
+export type InteractionTypes = "PING" | "APPLICATION_COMMAND" | "MESSAGE_COMPONENT"
+
+export type ApplicationCommandPermissionsTypes = "ROLE" | "USER";
+
+export class Base {
+	public readonly client: Client;
+}
+
+export type ApplicationCommandPermissionsFetchOptions = {
+  guildId?: Snowflake;
+  commandId?: Snowflake;
+};
+
+export type ApplicationCommandPermissionsSetOptions = ApplicationCommandPermissionsFetchOptions & {
+  permissions: ApplicationCommandPermissions[];
+  fullPermissions: FullApplicationCommandPermissions[];
+};
+
+export type ApplicationCommandPermissions = {
+  id: Snowflake;
+  type: ApplicationCommandPermissionsTypes;
+  permission: boolean;
+};
+
+export type ChannelTypes = "GUILD_TEXT" | "DM" | "GUILD_VOICE" | "GROUP_DM" | "GUILD_CATEGORY" | "GUILD_NEWS" | "GUILD_STORE" | "GUILD_NEWS_THREAD" | "GUILD_PUBLIC_THREAD" | "GUILD_PRIVATE_THREAD" | "GUILD_STAGE_VOICE";
+
+export type ApplicationCommandOptionsChoices = {
+  name: string;
+  value: string | number;
+};
+
+export type ApplicationCommandOptions = {
+  autocomplete?: boolean;
+  channel_types?: ChannelTypes[];
+  choices?: ApplicationCommandOptionsChoices[]
+};
+
+export type ApplicationCommandOptionsTypes = "SUB_COMMAND" | "SUB_COMMAND_GROUP" | "STRING" | "INTEGER" | "BOOLEAN" | "USER" | "CHANNEL" | "ROLE" | "MENTIONABLE" | "NUMBER";
+
+export type ApplicationCommandData = {
+  applicationId: Snowflake;
+  default_permission?: boolean;
+  guildId?: Snowflake;
+  id: Snowflake;
+  options?: ApplicationCommandOptions[];
+  description: string;
+  name: string;
+  required?: boolean;
+  type: ApplicationCommandOptionsTypes;
+};
+
+export type ApplicationCommandResolvable = ApplicationCommand | Snowflake;
+
+export type ApplicationCommandPermissionsAddOptions = {
+  guildId?: Snowflake;
+  commandId: Snowflake;
+  permissions: ApplicationCommandPermissions[];
+};
+
+export type ApplicationCommandPermissionsRemoveOptions = {
+  guildId?: Snowflake;
+  commandId: Snowflake;
+  users?: Snowflake[];
+  roles?: Snowflake[];
+};
+
+export type ApplicationCommandPermissionsHasOptions = {
+  guildId?: Snowflake;
+  commandId: Snowflake;
+  id: Snowflake;
+};
+
+export type ApplicationCommandPermissionsCommandPathOptions = {
+  commandId: Snowflake,
+  guildId?: Snowflake
+};
+
+export type ApplicationCommandFetchOptions = {
+  commandId: Snowflake;
+  guildId?: Snowflake;
+  cache?: boolean;
+  force: boolean;
+}
+
+export class ApplicationCommandManager extends BaseManager<Snowflake, {},ApplicationCommandResolvable> {
+  protected constructor(client: Client, iterable: Array<"Structure">, holds: unknown);
+  get permissions(): ApplicationCommandPermissionsManager;
+  add(data: ApplicationCommandData, cache:boolean, guildId: unknown): {}; 
```

Comment:
こんなのはどうでしょうか。
```suggestion
declare class OverridableManager<V, K, R> extends BaseManager<V, K, R> {
  public add(...args: any[]): any;
}

export class ApplicationCommandManager extends OverridableManager<Snowflake, ApplicationCommand, ApplicationCommandResolvable> {
  protected constructor(client: Client, iterable: Array<any>);
  get permissions(): ApplicationCommandPermissionsManager;
  add(data: ApplicationCommandData, cache:boolean, guildId: Snowflake): ApplicationCommand; 
```

変更内容
・https://github.com/discordjs/discord.js/blob/v12/typings/index.d.ts#L1982-L1985 を真似て、OverridableManagerを作ってそれを継承するようにしてみました。
・`ApplicationCommandManager` の親クラスを `OverridableManager<Snowflake, ApplicationCommand, ApplicationCommandResolvable>` に変更しました。
・コンストラクーの第２引数 `iterable` の型を `Array<"Structure">` から `Array<any>` に変更しました。
・コンストラクターの第３引数 `holds` が未使用であり、かつ、不必要だと判断したため、削除しました。
・ApplicationCommandManager#addの戻り値の型を `{}` から `ApplicationCommand` に変更しました。

Reply:
有能